webcommon/javascript2.model/src/org/netbeans/modules/javascript2/model/api/ModelUtils.java

```diff
@@ -1173,75 +1177,84 @@ public static boolean hasDeclaredProperty(JsObject jsObject) {
-                    String partfqn = props.toString();
-                    if (!partfqn.trim().equals("")) {
-                        String[] tokens = partfqn.split("\\.");
-                        for (int i = 0; i < tokens.length; i++) {
-                            object = ModelUtils.searchJsObjectByName(parent, newObjectName);
-                            for (JsObject prop : object.getProperties().values()) {
-                                if ((prop.getName().equals(tokens[i])) && (prop.isDeclared())) {
-                                    if ((prop.getAssignmentCount() > 0) && (deepCRT < MAX_RECURSION_DEEP_RESOLVING_ASSIGNMENTS)) {
-                                        for (TypeUsage type1 : prop.getAssignments()) {
-                                            return resolveTypes(object, String.join(".", Arrays.copyOfRange(tokens, i, tokens.length)), offset);
-                                        }
-                                    } else {
-                                        object = prop;
-                                        newObjectName = newObjectName + "." + prop.getName();
-                                        if (i == tokens.length - 1) {
-                                            typeResolved = true;
+        List<TypeUsage> diff = localResolved.stream().filter(type -> !alreadyResolved.contains(type)).collect(Collectors.toList());
+        if (diff.size() > 0) {
+            alreadyResolved.addAll(diff);
+            boolean typeResolved = false;
+            for (TypeUsage type : localResolved) {
+                if (type.isResolved()) {
+                    String newObjectName = type.getType();
+                    JsObject object = ModelUtils.searchJsObjectByName(parent, newObjectName);
+                    if ((object != null) && (object != parent)) {
+                        String partfqn = props.toString();
+                        if (!partfqn.trim().equals("")) {
+                            String[] tokens = partfqn.split("\\.");
+                            for (int i = 0; i < tokens.length; i++) {
+                                object = ModelUtils.searchJsObjectByName(parent, newObjectName);
+                                for (JsObject prop : object.getProperties().values()) {
+                                    if ((prop.getName().equals(tokens[i])) && (prop.isDeclared())) {
+                                        if (prop.getAssignmentCount() > 0) {
+                                            if (invokeCount == MAX_RECURSION_DEEP_RESOLVING_ASSIGNMENTS) {
+                                                LOG.log(Level.WARNING, "StackOverFlowError : {0} : {1}",
+                                                        new Object[]{object.getFullyQualifiedName(),
+                                                            object.getFileObject()});
+                                            }
+                                            for (TypeUsage type1 : prop.getAssignments()) {
```

Comment:
Need this check? (Here is within the `if (prop.getAssignmentCount() > 0)` block and `type1` is not used.)
If need it, should write like this?
```java
if (!prop.getAssignments().isEmpty()) {
...
}
```